solidity/test/vault/DonationVault.test.ts

```diff
@@ -0,0 +1,283 @@
+          await bank
+            .connect(account1)
+            .increaseBalanceAllowance(vault.address, 999)
+        })
+
+        after(async () => {
+          await restoreSnapshot()
+        })
+
+        it("should revert", async () => {
+          await expect(vault.connect(account1).donate(1000)).to.be.revertedWith(
+            "Transfer amount exceeds allowance"
+          )
+        })
+      }
+    )
+
+    context("when called with correct parameters", () => {
+      let tx: ContractTransaction
+
+      before(async () => {
+        await createSnapshot()
+
+        await bank.connect(bridge).increaseBalance(account1.address, 1000)
+        await bank
+          .connect(account1)
+          .increaseBalanceAllowance(vault.address, 1000)
+
+        tx = await vault.connect(account1).donate(1000)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should decrease donator's balance", async () => {
+        expect(await bank.balanceOf(account1.address)).to.be.equal(0)
+      })
+
+      it("should not increase vault's balance", async () => {
+        expect(await bank.balanceOf(vault.address)).to.be.equal(0)
+      })
+
+      it("should emit BalanceDecreased event", async () => {
+        await expect(tx)
+          .to.emit(bank, "BalanceDecreased")
+          .withArgs(vault.address, 1000)
+      })
+
+      it("should emit DonationReceived event", async () => {
+        await expect(tx)
+          .to.emit(vault, "DonationReceived")
+          .withArgs(account1.address, 1000)
+      })
+    })
+  })
+
+  describe("receiveBalanceApproval", () => {
+    context("when called not by the bank", () => {
+      it("should revert", async () => {
+        await expect(
+          vault.connect(bridge).receiveBalanceApproval(account1.address, 1000)
+        ).to.be.revertedWith("Caller is not the Bank")
+      })
+    })
+
+    context("when caller has not enough balance in the bank", () => {
+      before(async () => {
+        await createSnapshot()
+
+        await bank.connect(bridge).increaseBalance(account1.address, 999)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should revert", async () => {
+        await expect(
+          bank.connect(account1).approveBalanceAndCall(vault.address, 1000)
+        ).to.be.revertedWith("Amount exceeds balance in the bank")
+      })
+    })
+
+    context("when called with correct parameters", () => {
+      let tx: ContractTransaction
+
+      before(async () => {
+        await createSnapshot()
+
+        await bank.connect(bridge).increaseBalance(account1.address, 1000)
+
+        tx = await bank
+          .connect(account1)
+          .approveBalanceAndCall(vault.address, 1000)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should decrease donator's balance", async () => {
+        expect(await bank.balanceOf(account1.address)).to.be.equal(0)
```

Comment:
Can we `increaseBalance` for a bit more than 1000 to confirm just the amount from `approveBalanceAndCall` is donated?